### ISSUE TITLE
ixiatg-op-system has intermittent kne deploy health check failure 

### ISSUE THREAD

**Issue by kmo7** [openconfig/kne#466]
Alex Bortok asked me to create an issue for this which I reported to him today. It is happening on the 0.0.1-4557 no-license required ixia release.  May/may not happen on other releases.

+ kne deploy deploy/kne/kind-bridge.yaml command leads to below intermittently:
I1201 10:50:18.795513    9726 deploy.go:1283] Waiting on deployment "ixiatg-op-system" to be healthy
Error: failed to check if controller is healthy: context canceled before "ixiatg-op-system" healthy

+++ On the deployment VM:
oot@controller:~/kne# git log -1
commit c8895cb422754d7cb2e9847a1d8f74d6a60c72dd (HEAD -> main, origin/main, origin/HEAD)
Author: jasdeep-hundal [hundal@google.com](mailto:hundal@google.com)
Date:   Thu Nov 30 14:51:40 2023 -0800
 
    Update KENG configmap for release 0.1.0-81 (#465)
root@controller:~/kne#
 
root@controller:~/kne# cat /root/kne/manifests/keysight/ixiatg-configmap.yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: ixiatg-release-config
  namespace: ixiatg-op-system
data:
  versions: |
    {
        "release": "0.0.1-9999",
        "images": [
            {
                "name": "controller",
                "path": "ghcr.io/open-traffic-generator/keng-controller",
                "tag": "0.1.0-81"
            },
            {
                "name": "gnmi-server",
                "path": "ghcr.io/open-traffic-generator/otg-gnmi-server",
                "tag": "1.13.2"
            },
            {
                "name": "traffic-engine",
                "path": "ghcr.io/open-traffic-generator/ixia-c-traffic-engine",
                "tag": "1.6.0.100"
            },
            {
                "name": "protocol-engine",
                "path": "ghcr.io/open-traffic-generator/ixia-c-protocol-engine",
                "tag": "1.00.0.339"
            },
            {
                "name": "ixhw-server",
                "path": "ghcr.io/open-traffic-generator/keng-layer23-hw-server",
                "tag": "0.13.2-2"
            }
        ]
    }

+++ The loaded IXIA containers:
root@controller:~/kne# docker image ls
REPOSITORY                                                       TAG                          IMAGE ID       CREATED         SIZE
ghcr.io/open-traffic-generator/licensed/ixia-c-controller        0.0.1-4557                   f243c3b3c3f6   3 days ago      189MB
ncptx                                                            22.3X80-202310010340.0-EVO   adf87a0a65ca   2 months ago    5.06GB
ncptx                                                            latest                       adf87a0a65ca   2 months ago    5.06GB
ghcr.io/open-traffic-generator/ixia-c-traffic-engine             1.6.0.85                     649b50fbb3a8   2 months ago    270MB
ghcr.io/open-traffic-generator/licensed/ixia-c-protocol-engine   1.00.0.331                   1292a35c6102   2 months ago    447MB
ghcr.io/open-traffic-generator/ixia-c-gnmi-server                1.12.7                       68405a8c5d99   2 months ago    170MB
kindest/node                                                     v1.26.0                      6d3fbfb3da60   11 months ago   931MB

+++ The actual ixiatg-configmap.yaml for the 4557 release (which is not in the KNE repo -- as expected):
cat /root/ixiatg-configmap.yaml 
apiVersion: v1
kind: ConfigMap
metadata:
    name: ixiatg-release-config
    namespace: ixiatg-op-system
data:
    versions: |
        {
          "release": "0.0.1-4557",
          "images": [
                {
                    "name": "controller",
                    "path": "ghcr.io/open-traffic-generator/licensed/ixia-c-controller",
                    "tag": "0.0.1-4557"
                },
                {
                    "name": "gnmi-server",
                    "path": "ghcr.io/open-traffic-generator/ixia-c-gnmi-server",
                    "tag": "1.12.7"
                },
                {
                    "name": "traffic-engine",
                    "path": "ghcr.io/open-traffic-generator/ixia-c-traffic-engine",
                    "tag": "1.6.0.85"
                },
                {
                    "name": "protocol-engine",
                    "path": "ghcr.io/open-traffic-generator/licensed/ixia-c-protocol-engine",
                    "tag": "1.00.0.331"
                },
                {
                    "name": "ixhw-server",
                    "path": "ghcr.io/open-traffic-generator/ixia-c-ixhw-server",
                    "tag": "0.12.5-1"
                }
            ]
        }

**Comment by alexmasi** [comment]
`kne deploy` calls this func as a health check: https://github.com/openconfig/kne/blob/7ecf53ec2711e0c4d6f482a782064148344ff482/deploy/deploy.go#L1295

This is essentially confirming that the `ixiatg-op-system` deployment comes up healthy with the expected number of replicas.

The manifest outlining the ixia operator deployment is defined from keysight: https://github.com/openconfig/kne/blob/main/manifests/keysight/ixiatg-operator.yaml

Can you provide more information using `kubectl` to get deployment information when the failure occurs?

@anjan-keysight

**Comment by kmo7** [comment]
Hi Alex,

This issue is intermittent and so far we have not been able to reproduce it manually.
In automation runs, our KNE VM is terminated after the run ends. If you tell me exactly what other logs  you like to see, we can see if we can quarantine the system with the VM intact and collect them.

So far only seen this on the 4557 IXIA image. As noted in the case, the KNE repo’s ixiatg-configmap.yaml has moved  the later “license required” IXIA images.

Here is the full kne deply log FWIW. Had seen the func you mentioned. What is causing the health check cancellation? A timeout or?


+ kne deploy deploy/kne/kind-bridge.yaml

I1201 10:44:21.628213    9726 deploy.go:248] Deploying cluster...

I1201 10:44:21.657371    9726 deploy.go:534] kind version valid: got v0.17.0 want v0.17.0

I1201 10:44:21.657638    9726 deploy.go:545] Attempting to recycle existing cluster "kne"...

W1201 10:44:21.894113    9726 deploy.go:64] (kubectl): W1201 10:44:21.893741    9747 loader.go:222] Config not found: /etc/kubernetes/admin.conf

W1201 10:44:21.894996    9726 deploy.go:64] (kubectl): error: context "kind-kne" does not exist

I1201 10:44:21.899159    9726 deploy.go:570] Creating kind cluster with: [create cluster --name kne --image kindest/node:v1.26.0 --config /root/kne/manifests/kind/config.yaml]

W1201 10:44:22.031421    9726 deploy.go:84] (kind): Creating cluster "kne" ...

W1201 10:44:22.031456    9726 deploy.go:84] (kind):  • Ensuring node image (kindest/node:v1.26.0) 🖼  ...

W1201 10:45:01.229403    9726 deploy.go:84] (kind):  ✓ Ensuring node image (kindest/node:v1.26.0) 🖼

W1201 10:45:01.229576    9726 deploy.go:84] (kind):  • Preparing nodes 📦   ...

W1201 10:45:03.609468    9726 deploy.go:84] (kind):  ✓ Preparing nodes 📦

W1201 10:45:03.662175    9726 deploy.go:84] (kind):  • Writing configuration 📜  ...

W1201 10:45:04.069565    9726 deploy.go:84] (kind):  ✓ Writing configuration 📜

W1201 10:45:04.069600    9726 deploy.go:84] (kind):  • Starting control-plane 🕹️  ...

W1201 10:45:19.482908    9726 deploy.go:84] (kind):  ✓ Starting control-plane 🕹️

W1201 10:45:19.483088    9726 deploy.go:84] (kind):  • Installing StorageClass 💾  ...

W1201 10:45:20.376517    9726 deploy.go:84] (kind):  ✓ Installing StorageClass 💾

W1201 10:45:20.733892    9726 deploy.go:84] (kind): Set kubectl context to "kind-kne"

W1201 10:45:20.733946    9726 deploy.go:84] (kind): You can now use your cluster with:

W1201 10:45:20.733955    9726 deploy.go:84] (kind): kubectl cluster-info --context kind-kne

W1201 10:45:20.733995    9726 deploy.go:84] (kind): Have a nice day! 👋

I1201 10:45:20.735665    9726 deploy.go:582] Deployed kind cluster: kne

I1201 10:45:20.735836    9726 deploy.go:605] Found manifest "/root/kne/manifests/kind/bridge.yaml"

I1201 10:45:21.077654    9726 deploy.go:61] (kubectl): clusterrole.rbac.authorization.k8s.io/kindnet created

I1201 10:45:21.081897    9726 deploy.go:61] (kubectl): clusterrolebinding.rbac.authorization.k8s.io/kindnet created

I1201 10:45:21.087813    9726 deploy.go:61] (kubectl): serviceaccount/kindnet created

I1201 10:45:21.098498    9726 deploy.go:61] (kubectl): daemonset.apps/kindnet created

I1201 10:45:21.105297    9726 deploy.go:252] Cluster deployed

I1201 10:45:21.186946    9726 deploy.go:61] (kubectl): [0;32mKubernetes control plane[0m is running at [0;33mhttps://127.0.0.1:34485[0m

I1201 10:45:21.186991    9726 deploy.go:61] (kubectl): [0;32mCoreDNS[0m is running at [0;33mhttps://127.0.0.1:34485/api/v1/namespaces/kube-system/services/kube-dns:dns/proxy[0m

I1201 10:45:21.187002    9726 deploy.go:61] (kubectl): To further debug and diagnose cluster problems, use 'kubectl cluster-info dump'.

I1201 10:45:21.191022    9726 deploy.go:256] Cluster healthy

I1201 10:45:21.197538    9726 deploy.go:267] Checking kubectl versions.

I1201 10:45:21.299418    9726 deploy.go:322] Deploying ingress...

I1201 10:45:21.299517    9726 deploy.go:864] Creating metallb namespace

I1201 10:45:21.299526    9726 deploy.go:883] Deploying MetalLB from: /root/kne/manifests/metallb/manifest.yaml

I1201 10:45:21.552258    9726 deploy.go:61] (kubectl): namespace/metallb-system created

I1201 10:45:21.568584    9726 deploy.go:61] (kubectl): customresourcedefinition.apiextensions.k8s.io/addresspools.metallb.io created

I1201 10:45:21.579104    9726 deploy.go:61] (kubectl): customresourcedefinition.apiextensions.k8s.io/bfdprofiles.metallb.io created

I1201 10:45:21.588001    9726 deploy.go:61] (kubectl): customresourcedefinition.apiextensions.k8s.io/bgpadvertisements.metallb.io created

I1201 10:45:21.600970    9726 deploy.go:61] (kubectl): customresourcedefinition.apiextensions.k8s.io/bgppeers.metallb.io created

I1201 10:45:21.607157    9726 deploy.go:61] (kubectl): customresourcedefinition.apiextensions.k8s.io/communities.metallb.io created

I1201 10:45:21.613703    9726 deploy.go:61] (kubectl): customresourcedefinition.apiextensions.k8s.io/ipaddresspools.metallb.io created

I1201 10:45:21.622924    9726 deploy.go:61] (kubectl): customresourcedefinition.apiextensions.k8s.io/l2advertisements.metallb.io created

I1201 10:45:21.631880    9726 deploy.go:61] (kubectl): serviceaccount/controller created

I1201 10:45:21.637233    9726 deploy.go:61] (kubectl): serviceaccount/speaker created

I1201 10:45:21.644303    9726 deploy.go:61] (kubectl): role.rbac.authorization.k8s.io/controller created

I1201 10:45:21.650904    9726 deploy.go:61] (kubectl): role.rbac.authorization.k8s.io/pod-lister created

I1201 10:45:21.655520    9726 deploy.go:61] (kubectl): clusterrole.rbac.authorization.k8s.io/metallb-system:controller created

I1201 10:45:21.659689    9726 deploy.go:61] (kubectl): clusterrole.rbac.authorization.k8s.io/metallb-system:speaker created

I1201 10:45:21.669806    9726 deploy.go:61] (kubectl): rolebinding.rbac.authorization.k8s.io/controller created

I1201 10:45:21.676768    9726 deploy.go:61] (kubectl): rolebinding.rbac.authorization.k8s.io/pod-lister created

I1201 10:45:21.682638    9726 deploy.go:61] (kubectl): clusterrolebinding.rbac.authorization.k8s.io/metallb-system:controller created

I1201 10:45:21.687837    9726 deploy.go:61] (kubectl): clusterrolebinding.rbac.authorization.k8s.io/metallb-system:speaker created

I1201 10:45:21.694648    9726 deploy.go:61] (kubectl): secret/webhook-server-cert created

I1201 10:45:21.704458    9726 deploy.go:61] (kubectl): service/webhook-service created

I1201 10:45:21.715338    9726 deploy.go:61] (kubectl): deployment.apps/controller created

I1201 10:45:21.725142    9726 deploy.go:61] (kubectl): daemonset.apps/speaker created

I1201 10:45:21.738642    9726 deploy.go:61] (kubectl): validatingwebhookconfiguration.admissionregistration.k8s.io/metallb-webhook-configuration created

I1201 10:45:21.748514    9726 deploy.go:888] Creating metallb secret

I1201 10:45:21.751753    9726 deploy.go:1283] Waiting on deployment "metallb-system" to be healthy

I1201 10:50:11.602715    9726 deploy.go:1310] Deployment "metallb-system" healthy

I1201 10:50:11.605603    9726 deploy.go:912] Applying metallb ingress config

W1201 10:50:11.612660    9726 deploy.go:950] Failed to create address polling (will retry 5 times)

I1201 10:50:16.654067    9726 deploy.go:1283] Waiting on deployment "metallb-system" to be healthy

I1201 10:50:16.656854    9726 deploy.go:1310] Deployment "metallb-system" healthy

I1201 10:50:16.656897    9726 deploy.go:331] Ingress healthy

I1201 10:50:16.656922    9726 deploy.go:332] Deploying CNI...

I1201 10:50:16.656938    9726 deploy.go:1013] Deploying Meshnet from: /root/kne/manifests/meshnet/grpc/manifest.yaml

I1201 10:50:17.545340    9726 deploy.go:61] (kubectl): namespace/meshnet created

I1201 10:50:17.554004    9726 deploy.go:61] (kubectl): customresourcedefinition.apiextensions.k8s.io/gwirekobjs.networkop.co.uk created

I1201 10:50:17.563257    9726 deploy.go:61] (kubectl): customresourcedefinition.apiextensions.k8s.io/topologies.networkop.co.uk created

I1201 10:50:17.581230    9726 deploy.go:61] (kubectl): serviceaccount/meshnet created

I1201 10:50:17.585670    9726 deploy.go:61] (kubectl): clusterrole.rbac.authorization.k8s.io/meshnet-clusterrole created

I1201 10:50:17.590640    9726 deploy.go:61] (kubectl): clusterrolebinding.rbac.authorization.k8s.io/meshnet-clusterrolebinding created

W1201 10:50:17.601790    9726 deploy.go:64] (kubectl): Warning: spec.template.spec.nodeSelector[beta.kubernetes.io/arch]: deprecated since v1.14; use "kubernetes.io/arch" instead

I1201 10:50:17.602068    9726 deploy.go:61] (kubectl): daemonset.apps/meshnet created

I1201 10:50:17.607307    9726 deploy.go:1017] Meshnet Deployed

I1201 10:50:17.607355    9726 deploy.go:1022] Waiting on Meshnet to be Healthy

I1201 10:50:17.610719    9726 deploy.go:1043] Meshnet Healthy

I1201 10:50:17.610778    9726 deploy.go:342] CNI healthy

I1201 10:50:17.610813    9726 deploy.go:344] Deploying controller...

I1201 10:50:17.610850    9726 deploy.go:1236] Deploying IxiaTG controller from: /root/kne/manifests/keysight/ixiatg-operator.yaml

I1201 10:50:17.895522    9726 deploy.go:61] (kubectl): namespace/ixiatg-op-system created

I1201 10:50:17.905535    9726 deploy.go:61] (kubectl): customresourcedefinition.apiextensions.k8s.io/ixiatgs.network.keysight.com created

I1201 10:50:17.912051    9726 deploy.go:61] (kubectl): serviceaccount/ixiatg-op-controller-manager created

I1201 10:50:17.918837    9726 deploy.go:61] (kubectl): role.rbac.authorization.k8s.io/ixiatg-op-leader-election-role created

I1201 10:50:17.923206    9726 deploy.go:61] (kubectl): clusterrole.rbac.authorization.k8s.io/ixiatg-op-manager-role created

I1201 10:50:17.926670    9726 deploy.go:61] (kubectl): clusterrole.rbac.authorization.k8s.io/ixiatg-op-metrics-reader created

I1201 10:50:17.930368    9726 deploy.go:61] (kubectl): clusterrole.rbac.authorization.k8s.io/ixiatg-op-proxy-role created

I1201 10:50:17.937163    9726 deploy.go:61] (kubectl): rolebinding.rbac.authorization.k8s.io/ixiatg-op-leader-election-rolebinding created

I1201 10:50:17.941222    9726 deploy.go:61] (kubectl): clusterrolebinding.rbac.authorization.k8s.io/ixiatg-op-manager-rolebinding created

I1201 10:50:17.945794    9726 deploy.go:61] (kubectl): clusterrolebinding.rbac.authorization.k8s.io/ixiatg-op-proxy-rolebinding created

I1201 10:50:17.951488    9726 deploy.go:61] (kubectl): configmap/ixiatg-op-manager-config created

I1201 10:50:17.963030    9726 deploy.go:61] (kubectl): service/ixiatg-op-controller-manager-metrics-service created

I1201 10:50:17.972379    9726 deploy.go:61] (kubectl): deployment.apps/ixiatg-op-controller-manager created

I1201 10:50:17.977217    9726 deploy.go:1270] Deploying IxiaTG config map from: /root/kne/manifests/keysight/ixiatg-configmap.yaml

I1201 10:50:18.789481    9726 deploy.go:61] (kubectl): configmap/ixiatg-release-config created

I1201 10:50:18.795052    9726 deploy.go:1274] IxiaTG controller deployed

I1201 10:50:18.795513    9726 deploy.go:1283] Waiting on deployment "ixiatg-op-system" to be healthy

Error: failed to check if controller is healthy: context canceled before "ixiatg-op-system" healthy


Regards,
Kaveh



Juniper Business Use Only

**Comment by anjan-keysight** [comment]
Firstly I don't think this is related to the 0.0.1-4557 release (more precisely the keng-operator release 0.3.6).
This could be some timeout related issue. We can log response for "kubectl get pods -n ixiatg-op-system", just before health check.

**Comment by alexmasi** [comment]
I agree with Anjan, a `kubectl describe` on the operator pod may also be useful.

We also see flakes with operators coming up very very rarely and is most of the time just a timeout, we can consider increasing the wait time on the health check. It would be good to see what exactly is going wrong here before increasing the timeout.

**Comment by kmo7** [comment]
Thank you Alex Masi and all.
I have added the –progress flag to kne deploy which shows a lot of internal details of controllers coming up. See around 20 secs for the IXIA controller to pass health check from when it is spawned on my manual setup but this could get higher on loaded automation setups (we will see). Also saw some kne deploy fails on other controllers in earlier logs, so I also think it is not related to IXIA at this point (we will confirm based on more data).

Alex M. mentioned he will look into designing in more debuggability info for kne failures, but we have more info to go by now. I added  the following commands in case of kne deploy failure, but did not change kne repo internally.

   kubectl get pods -A -o wide
    kubectl get services -A
    kubectl describe pods -A

Regards,
Kaveh



Juniper Business Use Only
